swap/swap.go

```diff
@@ -419,6 +421,103 @@ func (s *Swap) Balances() (map[enode.ID]int64, error) {
 	return balances, nil
 }
 
+// Cheques returns all known last sent and received cheques, grouped by peer
+func (s *Swap) Cheques() (map[enode.ID]map[string]*Cheque, error) {
+	cheques := make(map[enode.ID]map[string]*Cheque)
+
+	// get peer cheques from memory
+	s.peersLock.Lock()
+	for peer, swapPeer := range s.peers {
+		swapPeer.lock.Lock()
+		cheques[peer] = make(map[string]*Cheque)
+		cheques[peer][lastSentChequeKey] = swapPeer.getLastSentCheque()
+		cheques[peer][lastReceivedChequeKey] = swapPeer.getLastReceivedCheque()
+		swapPeer.lock.Unlock()
+	}
+	s.peersLock.Unlock()
+
+	// add disk cheques for peers not already present
+	sentChequesIterFunction := func(key []byte, value []byte) (stop bool, err error) {
+		peer := keyToID(string(key), sentChequePrefix)
+		// make map if peer has no cheques entry yet
+		if _, peerHasCheques := cheques[peer]; !peerHasCheques {
+			cheques[peer] = make(map[string]*Cheque)
+		}
+		// add sent cheque from store
+		if peerSentCheque := cheques[peer][lastSentChequeKey]; peerSentCheque == nil {
+			var peerCheque Cheque
+			err = json.Unmarshal(value, &peerCheque)
+			if err == nil {
+				cheques[peer][lastSentChequeKey] = &peerCheque
+			}
+		}
+		// add nil as received cheque if not present
+		if _, peerHasReceivedCheque := cheques[peer][lastReceivedChequeKey]; !peerHasReceivedCheque {
+			cheques[peer][lastReceivedChequeKey] = nil
+		}
+		return stop, err
+	}
+	err := s.store.Iterate(sentChequePrefix, sentChequesIterFunction)
+	if err != nil {
+		return nil, err
+	}
+	receivedChequesIterFunction := func(key []byte, value []byte) (stop bool, err error) {
+		peer := keyToID(string(key), receivedChequePrefix)
+		// make map if peer has no cheques entry yet
+		if _, peerHasCheques := cheques[peer]; !peerHasCheques {
+			cheques[peer] = make(map[string]*Cheque)
+		}
+		// add received cheque from store
+		if peerReceivedCheque := cheques[peer][lastReceivedChequeKey]; peerReceivedCheque == nil {
+			var peerCheque Cheque
+			err = json.Unmarshal(value, &peerCheque)
+			if err == nil {
+				cheques[peer][lastReceivedChequeKey] = &peerCheque
+			}
+		}
+		// add nil as sent cheque if not present
+		if _, peerHasSentCheque := cheques[peer][lastSentChequeKey]; !peerHasSentCheque {
+			cheques[peer][lastSentChequeKey] = nil
+		}
+		return stop, err
+	}
+	err = s.store.Iterate(receivedChequePrefix, receivedChequesIterFunction)
+	if err != nil {
+		return nil, err
+	}
+
+	return cheques, nil
+}
+
+// PeerCheques returns the last sent and received cheques for a given peer
+func (s *Swap) PeerCheques(peer enode.ID) (map[string]*Cheque, error) {
```

Comment:
Based solely on the function description, this function could just be implemented by calling SentCheque and ReceivedCheque. Why did you choose not to re-use those functions, but implement it fully by yourself?

Reply:
see previous answer. it doesn't translate 100% since we are not iterating through the store here, but the idea is the same.

Reply:
I see why you can't reuse the other functions but I'm not sure I like the behaviour of this call. As far is I understand it if a peer has never sent or received a cheque before it returns:
* with `lastSentChequeKey` nil and `lastSentReceivedKey` nil if the peer is connected
* with `state.ErrNotFound` if the peer is not connected

I don't think connection status should matter for this request.

We cannot detect wether we know a peer based on the cheques alone. If we want to return an error for unknown peers we would have to check for balance instead. Alternatively we could just return `nil` for both values. Then the function would simplify to this:
```go
func (s *Swap) PeerCheques(peer enode.ID) (cheques map[string]*Cheque, err error) {
	var sentCheque, receivedCheque *Cheque
	if sentCheque, err = s.SentCheque(peer); err != nil && err != state.ErrNotFound {
		return nil, err
	}
	if receivedCheque, err = s.ReceivedCheque(peer); err != nil && err != state.ErrNotFound {
		return nil, err
	}
	return map[string]*Cheque{lastSentChequeKey: sentCheque, lastReceivedChequeKey: receivedCheque}, nil
}
```

Reply:
i was trying to make this function behave in the same manner as the `Balances` call, which returns `state.ErrNotFound` in case the peer isn't in memory nor on disk.

i guess it's different here: as you point out, a valid SWAP peer could still have no cheques in either of those places while still being valid :thinking: 

per my initial point, i would like this call to return `state.ErrNotFound` when the peer is unknown. but i am not a fan of querying balances to know this when we are asking for cheques info.

therefore, i will go with returning `nil` as a minimum, unless there are objections.

thanks